### DISCUSSION TITLE
[5.0] Standardise Front End User Account Templates

### DISCUSSION THREAD

**Discussion by leevigraham** [craftcms/cms#14133] [Ideas]
Currently the front end user templates (and associated actions) seem inconsistent. It would be great if the implementation and template variables could be standardised.

Here's what I think could be an improvement:

**Create "forms" for all front end user account templates.**

These would just be Yii Models that wrap validation, errors and default values. For example a login template could start with:

```twig
{% set form = loginForm ?? create('craft\\models\\forms\\login') %}
```

From that point on checking for errors etc is standardised eg: `{{ form.getErrors('loginName') }}`. No more errors passed to the template as a variable (which may or may not be defined).

I can kinda already do this with register templates as the variable returned from the action is a user… eg:

```twig
{% set form = user ?? create('craft\\elements\\user') %}
```

In addition to errors model properties / functions like `attributeLabels()`, `getAttributeHints()` could be used to create input labels, hints etc. One use case for this is changing the login form `loginName` label when the general config `useEmailAsUsername` is set… eg 'Username' or 'Email' (if `useEmailAsUsername` is set to true).

**Document flash usage**

I find it hard to follow when to use flash errors / success in front end templates vs the provided `errorMessage` variables in templates. Some kind of documentation on this would help.

**Comment by piotrpog** [comment]
I like idea of accessing error messages as object properties, instead of just twig variables.

**Comment by mandrasch** [comment]
**Add information about localized routing**

What I also noticed was that it is assumed that the paths are the same for all sites, e.g. doing localized routing like my-site.com/register and my-site.com/de/registrieren for multi site is not covered in the tutorial https://craftcms.com/knowledge-base/front-end-user-accounts 

(There are multiple ways of achieving this like routes.php and `siteUrl()`, use of Singles, using multiple folders in templates/ like de/ and en/, etc. - each with own pros and cons)

Would be cool to have at least a hint for users in that tutorial as well on how to get started 👍
